airflow/operators/python.py

```diff
@@ -167,6 +167,34 @@ def execute(self, context: Dict):
         return branch
 
 
+def create_branch_join(branch_operator, *args, **kwargs):
+    """
+    Create a join task for a branching logic. This join task is always executed regardless
+    of which branches are followed. It is only skipped if the ``branch_operator`` is skipped.
+    """
+    def python_callable(ti, **_):
+        from airflow.utils.session import create_session
+        from airflow.exceptions import AirflowSkipException
+        from airflow.utils.state import State
+        from airflow.models import TaskInstance
+
+        with create_session() as session:
+            branch_ti = session.query(TaskInstance).filter(
+                TaskInstance.dag_id == ti.dag_id,
+                TaskInstance.task_id == branch_operator.task_id,
+                TaskInstance.execution_date == ti.execution_date
+            ).one_or_none()
+
+            if not branch_ti:
+                return
+
+            if branch_ti.state == State.SKIPPED:
+                raise AirflowSkipException(f"Skipping because parent task {branch_operator.task_id} "
+                                           "is skipped.")
```

Comment:
task/ti shouldn't be setting state in this manner but not sure what other option we have over here !

Would love to know opinions from other committers @potiuk @mik-laj @ashb @feluelle @BasPH

Reply:
Thanks @kaxil. This is raising `AirflowSkipException` to cause the join task itself to be skipped. Please let me know if there are better ways to do this.

For other reviewers who haven't seen the previous discussion on the ancestor PR. What this `create_branch_join` does here is to make a join task like `join_2` in the following DAG. It makes sure the workflow can still work as we intuitively expect after the behaviour or `none_failed` is fixed by this PR. Logically, `join_2` should be skipped because its parent `BranchPythonOperator` `branch_2` has been skipped. Without `create_branch_join`, the `none_failed` task would actually become `success`.

![nested_branching](https://user-images.githubusercontent.com/6637585/75204369-a5bef200-57ab-11ea-9309-f3dc1cc044d4.png)

Reply:
Hi, @kaxil I've done the renaming you suggested to `create_branch_join_task()`. 

Any suggestion about how to make `join_2` skipped other than raising `AirflowSkipException`?

Reply:
What do you think of having 2 trigger rules? 

1) Fixed "none_failed"
2) none_failed_or_skipped

(2) can be used to get current "non_failed" like feature, isn't it?

Reply:
WDYT?

Reply:
Hi @kaxil  sorry for the late reply. Yes I think your suggestion sounds good. I'll think a bit more and work on fixing none_failed, and adding none_failed_or_skipped.

Reply:
@kaxil I have done the following. Could you take a second look? There are some elastic search related tests failing in Travis. Does not look related to my change.

1. Fixed "none_failed"
2. none_failed_or_skipped

Reply:
Sure I will take a look today, can you rebase on the latest master please to fix the Elasticsearch issue